firedrake/cofunction.py

```diff
@@ -79,6 +79,21 @@ def __del__(self):
         if hasattr(self, "_comm"):
             mpi.decref(self._comm)
 
+    def __call__(self, expr):
+        r"""Return the symbolic duality pairing, this is to evaluate this
+        :class:`.Cofunction` at an expression in the primal space, equivalent to
+        taking the dot product of the vectors of coefficients of the dual and
+        primal arguments.
+
+        :arg expr: a UFL expression to be paired with this :class:`.Cofunction`.
+
+        :returns: the symbolic duality pairing.
+        """
+        try:
+            return ufl.action(self, expr)
+        except TypeError:
+            return ufl.action(self, ufl.Interpolate(expr, self.function_space().dual()))
```

Comment:
A couple of comments:
- I think we should separate the concern of the function space from the actual duality pairing. I would advocate for having `expr` as a `Cofunction` on the right space, and then `__call__` should just return the action. If one wants to interpolate it should be done on the outside, e.g.:

```
c(Interpolate(f, V2))
```

where `c` lives in `V2*` and `f` in `V1`. This would be needed anyway if `f` is not a `Function` but an expression.

- `ufl.Interpolate` is also not usable at the moment since that just handles the symbolic part (e.g. differentiation), not the actual assembly which will be required whenever you need to evaluate that duality pairing. This will come with #2297.